test/e2e_node/container_lifecycle_pod_construction.go

```diff
@@ -124,28 +124,34 @@ func (o containerOutputList) String() string {
 
 // RunTogether returns an error the lhs and rhs run together
 func (o containerOutputList) RunTogether(lhs, rhs string) error {
-	lhsStart := o.findIndex(lhs, "Started", 0)
-	rhsStart := o.findIndex(rhs, "Started", lhsStart+1)
-
-	lhsFinish := o.findIndex(lhs, "Exiting", lhsStart+1)
-	rhsFinish := o.findIndex(rhs, "Exiting", rhsStart+1)
+	var startIdx int
+	for {
+		lhsStart := o.findIndex(lhs, "Started", startIdx)
+		// prepare for the next iteration
+		startIdx = lhsStart + 1
+		rhsStart := o.findIndex(rhs, "Started", lhsStart+1)
```

Comment:
This assumes that lhs *always* starts before rhs. Can you point me out why do you think it does?

Reply:
because it was intended like that by @tzneal and it makes sense for init containers (which sidecars are)

Reply:
So for these init containers:
```
				RestartPolicy: v1.RestartPolicyNever,
				InitContainers: []v1.Container{
					{
						Name:  init1,
						Image: busyboxImage,
						Command: ExecCommand(init1, execCommand{
							Delay:    1,
							ExitCode: 0,
						}),
					},
					{
						Name:  restartableInit1,
						Image: busyboxImage,
						Command: ExecCommand(restartableInit1, execCommand{
							Delay:    600,
							ExitCode: 0,
						}),
						RestartPolicy: &containerRestartPolicyAlways,
					},
					{
						Name:  init2,
						Image: busyboxImage,
						Command: ExecCommand(init2, execCommand{
							Delay:    1,
							ExitCode: 0,
						}),
					},
					{
						Name:  restartableInit2,
						Image: busyboxImage,
						Command: ExecCommand(restartableInit2, execCommand{
							Delay:    600,
							ExitCode: 0,
						}),
						RestartPolicy: &containerRestartPolicyAlways,
					},
					{
						Name:  init3,
						Image: busyboxImage,
						Command: ExecCommand(init3, execCommand{
							Delay:    1,
							ExitCode: 0,
						}),
					},
				},
```
we expect `restartableInit2` to always start before `restartableInit2`. Can you explain why? They have the same Delay, so can start at approximately the same time, no?

Reply:
they should not start in parallel, but one after the other once the previous one has exited (non-sidecar) or is ready (restartable=sidecar)

Reply:
Makes sense. Thanks for the explanations!
However, I'm still wondering how this test case worked in the code in master branch, considering this:
- https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_test.go#L1168
- https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/container_lifecycle_pod_construction.go#L128